### PR TITLE
fix(mcp): Harden lifecycle teardown and observability

### LINKED ISSUE #273: MCP server processes orphaned when parent (Claude CLI) dies — 4GB memory leak
## Problem

When XcodeBuildMCP is launched as an MCP server via `npm exec xcodebuildmcp@latest mcp` (as Claude Code does), the process persists indefinitely after the parent process exits. Each orphaned instance consumes up to **4GB of resident memory**.

This is especially painful when running multiple Claude Code sessions — orphaned `xcodebuildmcp` processes accumulate and grind the system to a halt.

```
$ ps aux | grep xcodebuild
dale  24113 352.3 23.5 450778096 3945760  ??  R  3:59PM  1:15.03 node ...xcodebuildmcp mcp
dale  24114   0.0  0.3 446631664  48928   ??  S  3:59PM  0:00.84 node ...xcodebuildmcp mcp
dale  24053   0.0  0.3 435888672  47712   ??  S  3:59PM  0:00.63 npm exec xcodebuildmcp@latest mcp
dale  24052   0.0  0.3 435888944  47712   ??  S  3:59PM  0:00.64 npm exec xcodebuildmcp@latest mcp
```

## Root Cause

`start-mcp-server.js` relies on `stdin` end/close events to detect parent death (lines 99–104). However, when spawned through `npm exec`, there's an intermediate node wrapper process. When the parent (Claude CLI) dies:

1. The `npm exec` wrapper becomes orphaned (reparented to launchd/PID 1)
2. The wrapper doesn't close its stdin pipe to the actual xcodebuildmcp process
3. The MCP server never receives the stdin end/close event
4. The process runs forever with no work to do

Additionally, unlike daemon mode (which has `idle-shutdown.js` with a 10-minute timeout), the MCP server mode has **no idle timeout at all**.

## Suggested Fix

Add a parent-PID watchdog to `start-mcp-server.js`. When the parent process dies, macOS/Linux reparents the child to init/launchd (PID 1). Polling `process.ppid` detects this:

```diff
--- a/src/server/start-mcp-server.ts
+++ b/src/server/start-mcp-server.ts
@@ in startMcpServer(), after enrichSentryContext() and before shuttingDown declaration
+    // Parent-process watchdog: detect orphaned process when parent dies.
+    // When the parent (Claude CLI / npm exec) exits, the OS reparents us to
+    // init/launchd (PID 1). Poll every 5s and shut down if that happens.
+    const originalPpid = process.ppid;
+    const ppidWatchdog = setInterval(() => {
+      if (process.ppid !== originalPpid) {
+        clearInterval(ppidWatchdog);
+        log("info", `Parent process died (was ${originalPpid}, now ${process.ppid}); shutting down`);
+        void shutdown("parent-died");
+      }
+    }, 5000);
+    ppidWatchdog.unref(); // Don't keep the event loop alive just for this
```

This is 8 lines, zero dependencies, works on macOS and Linux, and has no effect during normal operation.

## Environment

- macOS 15 (Darwin 25.3.0), Apple Silicon
- XcodeBuildMCP 2.2.1
- Claude Code 2.1.75 (VS Code extension)
- Launched via: `npm exec xcodebuildmcp@latest mcp`

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by cameroncooke**
Redesign MCP shutdown so disconnect/crash teardown is bounded and deterministic while keeping high-value telemetry.

Issue #273 reported orphaned MCP processes and memory growth after client teardown. The core change is moving shutdown orchestration into a dedicated runner that enforces bounded cleanup and explicit exit ordering. This removes the old inline shutdown branching and prevents disconnect-time teardown from getting stuck in an unbounded graceful path.

Key updates in this branch:
- Add `src/server/mcp-shutdown.ts` as the canonical shutdown runner
- Add shutdown state gating for stdio suppression + Sentry capture sealing
- Enrich lifecycle snapshots/reasons (`ppid`, `orphaned`, `stderr-error`)
- Replace MCP shutdown use of `Sentry.close(...)` with bounded `Sentry.flush(...)`
  after a single shutdown summary event
- Add bounded bulk cleanup for tracked resources (sim/device log sessions,
  video capture sessions, Swift tracked processes, debugger sessions,
  watcher/bridge shutdown)
- Keep tool-facing behavior stable while reusing shared stop helpers
- Extend tests around lifecycle mapping, shutdown ordering, seal/suppression,
  and shutdown helper behavior
- Add reproducible lifecycle leak script improvements and a repro npm script

I considered keeping the previous "disconnect fast-exit only" containment, but it drops useful shutdown telemetry and leaves cleanup ownership fragmented. This implementation keeps shutdown bounded while preserving actionable end-of-life diagnostics.

This branch also includes the MCP SDK bump to `@modelcontextprotocol/sdk@^1.27.1`.

Fixes #273

Validation and repro evidence:
- Ran the focused Sentry teardown harness with Sentry enabled across
  `REPRO_DRAIN_MODE=close|flush|none` in both parent-hard-exit and graceful
  stdio-close modes.
- Ran wrapper/logging toggles on parent-hard-exit (`wrap` on/off,
  `enableLogs` on/off, `captureLogsToSentry` on/off).
- Ran the real server lifecycle repro script in both
  `--shutdown-mode parent-hard-exit` and `--shutdown-mode graceful-stdin`.

These repro runs provided direct evidence that the shutdown contract now exits
boundedly without lingering spawned MCP processes in the sampled windows, and
that flush-mode teardown no longer shows the previous runaway signature.

### CODE REVIEW

File: src/server/mcp-lifecycle.ts [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+  };
+  const handleStdoutError = (error: unknown): void => {
+    if (!isBrokenPipeLikeError(error)) {
+      return;
+    }
+    void coordinator.shutdown('stdout-error', error);
+  };
+  const handleUncaughtException = (error: unknown): void => {
+    void coordinator.shutdown('uncaught-exception', error);
+  };
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-lifecycle.ts:363]: Audited this path and we are not applying a code change here. The process handlers intentionally trigger `coordinator.shutdown()`, and process termination is controlled by the shutdown coordinator path (bounded cleanup plus explicit exit). We re-ran the lifecycle repro harness in `parent-hard-exit` and `graceful-stdin` modes and did not reproduce premature termination from this handler wiring.

File: src/mcp/tools/swift-package/active-processes.ts [hunk 2]
```diff
@@ -42,3 +46,100 @@ export const clearAllProcesses = (): void => {
...
+  try {
+    info.process.kill('SIGTERM');
+  } catch (error) {
+    return { error: error instanceof Error ? error.message : String(error) };
+  }
+
+  const alreadyExited = info.process.exitCode != null || info.process.killed === true;
+  if (alreadyExited) {
+    return {};
+  }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/mcp/tools/swift-package/active-processes.ts:72]: Fixed in f2e1f702. We no longer treat `ChildProcess.killed` as an exit signal. Exit detection now relies on `exitCode`/`signalCode`, and we added a regression test that verifies we still escalate from SIGTERM to SIGKILL when the child has not actually exited.

File: src/utils/video_capture.ts [hunk 3]
```diff
@@ -28,21 +28,93 @@ export interface AxeHelpers {
...
-        const child = sess.process as ChildProcess | undefined;
-        child?.kill?.('SIGINT');
-      } catch {
-        // ignore
-      } finally {
-        sess.releaseActivity?.();
-        sessions.delete(simulatorUuid);
-      }
+    for (const simulatorUuid of sessions.keys()) {
+      void stopSession(simulatorUuid, { timeoutMs: 250 });
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/video_capture.ts:117]: Audited this path and we are not applying a code change here. In this handler we delete the current key while iterating, which does not skip remaining entries in this usage. For graceful bulk shutdown, `stopAllVideoCaptureSessions` already snapshots keys with `Array.from(sessions.keys())` before iterating.

File: src/server/mcp-shutdown.ts [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+    ['simulator-logs.stop-all', () => stopAllLogCaptures(STEP_TIMEOUT_MS)],
+    ['device-logs.stop-all', () => stopAllDeviceLogCaptures(STEP_TIMEOUT_MS)],
+    ['video-capture.stop-all', () => stopAllVideoCaptureSessions(STEP_TIMEOUT_MS)],
+    ['swift-processes.stop-all', () => stopAllTrackedProcesses(STEP_TIMEOUT_MS)],
+  ];
+
+  for (const [name, operation] of cleanupSteps) {
+    const outcome = await runStep(name, STEP_TIMEOUT_MS, operation);
+    pushStep(name, outcome);
+  }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-shutdown.ts]: Fixed in 60ae62f5. Bulk cleanup step budgets are now scaled by tracked item count while keeping a bounded per-item timeout, so sequential stop-all loops are not prematurely timed out by the outer step timer. Added regression coverage in `src/server/__tests__/mcp-shutdown.test.ts`.

File: scripts/repro-mcp-lifecycle-leak.ts [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+    );
+  });
+}
+
+async function runParentHardExitIteration(closeDelayMs: number): Promise<IterationResult> {
+  return new Promise((resolve) => {
+    const helper = spawn(
+      process.execPath,
+      [
+        'scripts/repro-mcp-parent-exit-helper.mjs',
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply scripts/repro-mcp-lifecycle-leak.ts:206]: Fixed in 60ae62f5. Added and committed `scripts/repro-mcp-parent-exit-helper.mjs`, so `npm run repro:mcp-lifecycle-leak -- --shutdown-mode parent-hard-exit` now works from a clean checkout.

File: src/server/mcp-shutdown.ts [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+
+  try {
+    const timeoutPromise = new Promise<'timed_out'>((resolve) => {
+      timeoutHandle = createTimer(timeoutMs, () => resolve('timed_out'));
+    });
+    const operationPromise = operation();
+    const outcome = await Promise.race([
+      operationPromise.then((value) => ({ kind: 'value' as const, value })),
+      timeoutPromise.then(() => ({ kind: 'timed_out' as const })),
+    ]);
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-shutdown.ts]: Fixed in 60ae62f5. `runStep` now converts operation outcomes into a structured race result and consumes late rejections, preventing unhandled rejection noise if timeout wins first.

File: src/mcp/tools/logging/stop_device_log_cap.ts [hunk 7]
```diff
@@ -272,62 +163,43 @@ export async function stopDeviceLogCapture(
...
-
-  const logContentMatch = successText.match(/--- Captured Logs ---\n([\s\S]*)$/);
-  const logContent = logContentMatch?.[1] ?? '';
-
-  return { logContent };
+  return { logContent: result.logContent };
 }
 
-export const schema = stopDeviceLogCapSchema.shape; // MCP SDK compatibility
+export { stopAllDeviceLogCaptures };
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/mcp/tools/logging/stop_device_log_cap.ts:200]: Audited: no code change for this item. The re-export is currently intentional API surface and not causing correctness or shutdown regression risk in this PR.

File: src/utils/video_capture.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/video_capture.ts:117]: Audited: no code change for this item. In the `exit` handler, the synchronous part of `stopSession` still runs (session removal and SIGINT best-effort), while async completion work cannot be guaranteed at process-exit by design. We’re keeping deterministic shutdown ownership in `mcp-shutdown` where bounded cleanup is awaited before `process.exit(...)`.

File: src/server/mcp-lifecycle.ts [hunk 8]
```diff
@@ -0,0 +1,460 @@
...
+    void coordinator.shutdown('sigint');
+  };
+  const handleStdinEnd = (): void => {
+    suppressProcessStdioWrites();
+    void coordinator.shutdown('stdin-end');
+  };
+  const handleStdinClose = (): void => {
+    suppressProcessStdioWrites();
+    void coordinator.shutdown('stdin-close');
+  };
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-lifecycle.ts:346]: Audited: no code change for this item. Stdio suppression before disconnect shutdown is intentional to avoid writes into broken pipes during orphan/transport-loss paths. We still emit shutdown telemetry through the structured Sentry summary path.

File: src/server/mcp-shutdown.ts [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+    : DEFAULT_SERVER_CLOSE_TIMEOUT_MS;
+
+  const serverCloseOutcome = await runStep('server.close', serverCloseTimeout, async () => {
+    await input.server?.close();
+  });
+  pushStep('server.close', serverCloseOutcome);
+
+  const perItemTimeoutMs = STEP_TIMEOUT_MS;
+  const bulkStepTimeoutMs = (itemCount: number): number => {
+    return Math.max(STEP_TIMEOUT_MS, itemCount * perItemTimeoutMs);
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-shutdown.ts]: Fixed in 0cf824ba. Added explicit outer timeout headroom for bulk cleanup step budgets so one-item boundary cases do not race the same 1000ms inner timeout. Added regression tests covering one-item and multi-item cases.

File: src/server/mcp-shutdown.ts [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    {
+      name: 'xcode-tools-bridge.shutdown',
+      timeoutMs: STEP_TIMEOUT_MS,
+      operation: () => shutdownXcodeToolsBridge(),
+    },
+    {
+      name: 'debugger.dispose-all',
+      timeoutMs: STEP_TIMEOUT_MS,
+      operation: () => getDefaultDebuggerManager().disposeAll(),
+    },
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/server/mcp-shutdown.ts:196]: Fixed in 0cf824ba. `debugger.dispose-all` now uses a larger dynamic timeout budget based on active debugger session count, and we added regression coverage proving a >1000ms dispose path completes.

File: src/mcp/tools/swift-package/active-processes.ts [hunk 11]
```diff
@@ -1,44 +1,147 @@
...
+      errors.push(`${pid}: ${result.error}`);
+    }
+  }
+
+  return {
+    stoppedProcessCount: pids.length - errors.length,
+    errorCount: errors.length,
+    errors,
+  };
+}
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/mcp/tools/swift-package/active-processes.ts:147]: Valid — fixed in bd6c7f75. `terminateTrackedProcess` now returns `{ status: 'not-found', error: 'Process not found: ${pid}' }`, consistent with the sibling cleanup helpers.

File: src/utils/video_capture.ts [hunk 12]
```diff
@@ -28,21 +28,93 @@ export interface AxeHelpers {
   getBundledAxeEnvironment: () => Record<string, string>;
 }
 
+function createTimeoutPromise(timeoutMs: number): Promise<'timed_out'> {
+  return new Promise((resolve) => {
+    const timer = setTimeout(() => resolve('timed_out'), timeoutMs);
+    timer.unref?.();
+  });
+}
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/utils/video_capture.ts:36]: Out of scope for this PR. The duplication predates this branch (`log_capture.ts` and `video_capture.ts` copies existed before). Happy to extract a shared utility in a follow-up.

File: src/mcp/tools/swift-package/active-processes.ts [hunk 13]
```diff
@@ -1,44 +1,147 @@
...
+
+  const outcome = await Promise.race([exitPromise, createTimeoutPromise(timeoutMs)]);
+  if (outcome === 'timed_out') {
+    try {
+      info.process.kill('SIGKILL');
+      usedForceKill = true;
+    } catch {
+      // ignore
+    }
+  }
```
[the remark this answers is not in the document]
Comment (cameroncooke) [reply src/mcp/tools/swift-package/active-processes.ts:93]: Not a real leak. After the timeout, SIGKILL is sent immediately (line 88). SIGKILL is unconditional -- the OS terminates the process, Node.js receives SIGCHLD, and the `exit` event fires, resolving `exitPromise` and releasing the listener. The only scenario where SIGKILL doesn't work is a kernel D-state process, which no userspace cleanup can fix. This code also only runs during shutdown, so the Node.js process itself exits moments later.
